options/posix/generic/syslog-stubs.cpp

```diff
@@ -44,13 +66,80 @@ int setlogmask(int) {
 	__builtin_unreachable();
 }
 
-void syslog(int, const char *, ...) {
-	__ensure(!"Not implemented");
-	__builtin_unreachable();
+static int is_lost_conn(int e) {
+	return e == ECONNREFUSED || e == ECONNRESET || e == ENOTCONN || e == EPIPE;
 }
 
-void vsyslog(int, const char *, va_list) {
-	__ensure(!"Not implemented");
-	__builtin_unreachable();
+static void _vsyslog(int priority, const char *message, va_list ap) {
+	char timebuf[16];
+	time_t now;
+	struct tm tm;
+	char buf[1024];
+	int errno_save = errno;
+	int pid;
+	int l, l2;
+	int hlen;
+	int fd;
+
+	if(log_fd < 0)
+		__openlog();
+
+	if(use_mlibc_logger) {
+		vsnprintf(buf, sizeof buf, message, ap);
+		mlibc::infoLogger() << "mlibc: syslog: " << buf << frg::endlog;
+		return;
+	}
+
+	if(!(priority & LOG_FACMASK))
+		priority |= log_facility;
+
+	now = time(NULL);
+	gmtime_r(&now, &tm);
+	// HACK: %e is unsupported, replacing with a less then ideal alternative.
+	//strftime(timebuf, sizeof timebuf, "%b %e %T", &tm);
+	strftime(timebuf, sizeof timebuf, "%b %d %T", &tm);
+
+	pid = (log_opt & LOG_PID) ? getpid() : 0;
+	l = snprintf(buf, sizeof buf, "<%d>%s %n%s%s%.0d%s: ",
+		priority, timebuf, &hlen, log_ident, "[" + !pid, pid, "]" + !pid);
+	errno = errno_save;
+	l2 = vsnprintf(buf + l, sizeof buf - l, message, ap);
+	if(l2 >= 0) {
+		if(l2 >= sizeof buf - l)
+			l = sizeof buf - 1;
+		else
+			l += l2;
+		if(buf[l - 1] != '\n')
+			buf[l++] = '\n';
+		if(send(log_fd, buf, l, 0) < 0 && (!is_lost_conn(errno)
+		    || connect(log_fd, (const sockaddr *)&log_addr, sizeof log_addr) < 0
+		    || send(log_fd, buf, l, 0) < 0)
+		    && (log_opt & LOG_CONS)) {
+			fd = open("/dev/console", O_WRONLY|O_NOCTTY|O_CLOEXEC);
+			if(fd >= 0) {
+				dprintf(fd, "%.*s", l - hlen, buf + hlen);
+				close(fd);
+			}
+		}
+		if(log_opt & LOG_PERROR)
+			dprintf(2, "%.*s", l - hlen, buf + hlen);
+	}
+}
+
+void syslog(int priority, const char *format, ...) {
+	va_list ap;
+	va_start(ap, format);
+	vsyslog(priority, format, ap);
+	va_end(ap);
 }
 
+void vsyslog(int priority, const char *message, va_list ap) {
+	int cs;
+	if(!(log_mask & LOG_MASK(priority & 7)) || (priority & ~0x3ff))
+		return;
+	//pthread_setcancelstate(PTHREAD_CANCEL_DISABLE, &cs);
+	//LOCK(lock);
+	_vsyslog(priority, message, ap);
+	//UNLOCK(lock);
```

Comment:
The log can be implemented on top of `mlibc::FutexLock`, https://github.com/managarm/mlibc/blob/master/options/internal/include/mlibc/lock.hpp#L8.